### PR TITLE
Fix #3208: Tests added for get/patch/delete allowed_email_addresses functions of…

### LINKED ISSUE #3208: [Backend] Add tests for GET/PATCH/DELETE allowed_email_ids APIs
### Description

PR #3194 adds an API to get, append and delete email ids allowed to participate in a challenge phase. The PR is missing tests. We need to add tests to fetch, append a set of email ids and delete existing email ids from `allowed_email_ids` field.

### PR DESCRIPTION
**Pull request by GrimmeryGonsa**
This PR fixes #3208

### CODE REVIEW

**Comment by GrimmeryGonsa** [comment]
unable to identify the reason behind the build failure. is it due to visual indentation errors???

File: tests/unit/challenges/test_views.py [hunk 1]
```diff
@@ -4287,7 +4286,7 @@ def test_get_challenge_phases_by_challenge_pk(self):
                 "is_submission_public": self.challenge_phase.is_submission_public,
                 "codename": self.challenge_phase.codename,
                 "test_annotation": "http://testserver%s"
-                % (self.challenge_phase.test_annotation.url),
+                                   % (self.challenge_phase.test_annotation.url),
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4289]: @GrimmeryGonsa can you revert the indentation changes. Were this done by pre-commit?

File: tests/unit/challenges/test_views.py [hunk 2]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
 
         self.assertEqual(response.data, expected)
         self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+
+class TestAllowedEmailIds(BaseChallengePhaseClass):
+
+    def check_api_get_request(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4539]: rename method to `test_get_or_update_allowed_email_ids_success`

File: tests/unit/challenges/test_views.py [hunk 3]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
+            kwargs={"challenge_pk": self.challenge.pk,
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4542]: Please fix the indentation, look at other tests for reference

File: tests/unit/challenges/test_views.py [hunk 4]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+                    },
+        )
+        expected = [{
+            "allowed_email_ids": self.challenge_phase.allowed_email_ids,
+        }]
+        response = self.client.get(self.url, {}, format='json')
+        self.assertEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def check_api_patch_request(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4553]: Change method name to `test_get_or_update_allowed_email_ids_patch_success`

File: tests/unit/challenges/test_views.py [hunk 5]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
+            kwargs={"challenge_pk": self.challenge.pk,
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4556]: Please fix the indentation

File: tests/unit/challenges/test_views.py [hunk 6]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+                    "phase_pk": self.challenge_phase.pk,
+                    })
+        expected = ["abc@def.com", "sample@fortest.com"]
+        for ele in self.challenge_phase.allowed_email_ids:
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4560]: rename `ele` to `allowed_email_id`

File: tests/unit/challenges/test_views.py [hunk 7]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+            expected.append(ele)
+        dummy_data = ["abc@def.com", "sample@fortest.com"]
+        data = [{
+            "allowed_email_ids": dummy_data,
+        }]
+        response = self.client.patch(self.url, data)
+        self.assertCountEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def check_api_delete_request(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4570]: rename method to `test_get_or_update_allowed_email_ids_delete_request`

File: tests/unit/challenges/test_views.py [hunk 8]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
+            kwargs={"challenge_pk": self.challenge.pk,
+                    "phase_pk": self.challenge_phase.pk,
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4574]: Fix indentation

File: tests/unit/challenges/test_views.py [hunk 9]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+                    },
+        )
+        # add dummy data
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4577]: remove the comment

File: tests/unit/challenges/test_views.py [hunk 10]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        dummy_data = ["abc@def.com", "sample@fortest.com"]
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4578]: change email domains to `user1@example.com` and `user2@example.com`. Rename variable `dummy_data` to `allowed_email_ids`

File: tests/unit/challenges/test_views.py [hunk 11]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        data = [{
+            "allowed_email_ids": dummy_data,
+        }]
+        self.client.patch(self.url, data)
+        # actual test...here we check if api is able to delete the dummy data added before
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4583]: Remove comment

File: tests/unit/challenges/test_views.py [hunk 12]
```diff
@@ -4533,3 +4532,59 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+    def check_api_patch_request(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
+            kwargs={"challenge_pk": self.challenge.pk,
+                    "phase_pk": self.challenge_phase.pk,
+                    })
+        expected = ["abc@def.com", "sample@fortest.com"]
+        for ele in self.challenge_phase.allowed_email_ids:
+            expected.append(ele)
+        dummy_data = ["abc@def.com", "sample@fortest.com"]
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4562]: change email domains to `user1@example.com` and `user2@example.com`. Rename variable `dummy_data` to `allowed_email_ids`

**Review by Ram81 — CHANGES_REQUESTED** [review]
@GrimmeryGonsa you also need to add tests for cases where exceptions occur.

File: tests/unit/challenges/test_views.py (hunk 1 above)
Comment (GrimmeryGonsa) [reply tests/unit/challenges/test_views.py:4289]: initially my pull request wasnt getting accepted, so i used pycharm to correct indentation, its probably due to that only.
I will try to revert it...

**Comment by GrimmeryGonsa** [comment]
> @GrimmeryGonsa you also need to add tests for cases where exceptions occur.

i am facing some difficulty for creating test for {{ if serializer.is_valid() }} condition, as serializers are created dynamically when the update_allowed_email_ids method is called, furthermore the components of the serializer ie. {challange_phase, data and context} are tested separately, so is there any need for creating test for that?.
if it is necessary then please guide me how to do that

**Comment by GrimmeryGonsa** [comment]
Hey @Ram81 , can you could look into "test_get_or_update_allowed_email_ids_patch_success" method?
i don't know why but the test case are failing for that particular method ......
it would be great if you could lend a helping hand 😅

**Comment by GrimmeryGonsa** [comment]
@Ram81 it would be great if you could review it 😅...

File: tests/unit/challenges/test_views.py [hunk 13]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        expected.extend(self.challenge_phase.allowed_email_ids)
+        allowed_email_ids = ["user1@example.com", "user2@example.com"]
+        data = {
+            "allowed_email_ids": allowed_email_ids,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertCountEqual(response.data["allowed_email_ids"], expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_or_update_allowed_email_ids_delete_request(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4573]: Change `test_get_or_update_allowed_email_ids_delete_request` to `test_get_or_update_allowed_email_ids_delete_success`

File: tests/unit/challenges/test_views.py [hunk 14]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        }
+        self.client.patch(self.url, data)
+        expected = {
+            "allowed_email_ids": self.challenge_phase.allowed_email_ids,
+        }
+        response = self.client.delete(self.url, data)
+        self.assertCountEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_if_allowed_email_ids_is_not_list(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4593]: change to `test_update_allowed_email_ids_with_invalid_input`

File: tests/unit/challenges/test_views.py [hunk 15]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        )
+        allowed_email_ids = ('user1@example.com')
+        data = {
+            "allowed_email_ids": allowed_email_ids,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data["error"], "Field allowed_email_ids should be a list.")
+
+    def test_if_allowed_email_ids_is_none(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4609]: change to `test_update_allowed_email_ids_when_input_is_none`

File: tests/unit/challenges/test_views.py [hunk 16]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+            },
+        )
+        data = {
+            "allowed_email_ids": None,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data["error"], "Field allowed_email_ids is missing.")
+
+    def test_if_challenge_phase_does_not_exist(self):
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4624]: change to `test_get_allowed_email_ids_when_challenge_phase_does_not_exist`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: tests/unit/challenges/test_views.py [hunk 17]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
 
         self.assertEqual(response.data, expected)
         self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+
+class TestAllowedEmailIds(BaseChallengePhaseClass):
+
+    def test_get_or_update_allowed_email_ids_success(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4542]: @GrimmeryGonsa can you use double quotes here

File: tests/unit/challenges/test_views.py [hunk 18]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        expected = {
+            "allowed_email_ids": self.challenge_phase.allowed_email_ids,
+        }
+        response = self.client.get(self.url, {}, format='json')
+        self.assertEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_or_update_allowed_email_ids_patch_success(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4557]: same here

File: tests/unit/challenges/test_views.py [hunk 19]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        data = {
+            "allowed_email_ids": allowed_email_ids,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertCountEqual(response.data["allowed_email_ids"], expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_or_update_allowed_email_ids_delete_success(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4575]: same here

File: tests/unit/challenges/test_views.py [hunk 20]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        expected = {
+            "allowed_email_ids": self.challenge_phase.allowed_email_ids,
+        }
+        response = self.client.delete(self.url, data)
+        self.assertCountEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_update_allowed_email_ids_with_invalid_input(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4595]: same here

File: tests/unit/challenges/test_views.py [hunk 21]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        data = {
+            "allowed_email_ids": allowed_email_ids,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data["error"], "Field allowed_email_ids should be a list.")
+
+    def test_update_allowed_email_ids_when_input_is_none(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4611]: same here

File: tests/unit/challenges/test_views.py [hunk 22]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+        data = {
+            "allowed_email_ids": None,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data["error"], "Field allowed_email_ids is missing.")
+
+    def test_get_allowed_email_ids_when_challenge_phase_does_not_exist(self):
+        self.url = reverse_lazy(
+            'challenges:get_or_update_allowed_email_ids',
```
Comment (Ram81) [inline tests/unit/challenges/test_views.py:4626]: same here

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: tests/unit/challenges/test_views.py (hunk 22 above)
Comment (GrimmeryGonsa) [reply tests/unit/challenges/test_views.py:4626]: ah... sorry for that

**Review by Ram81 — APPROVED** [review]
LGTM 👍. Nice work @GrimmeryGonsa

**Comment by GrimmeryGonsa** [comment]
> LGTM 👍. Nice work @GrimmeryGonsa

Thank you....😁😁

File: tests/unit/challenges/test_views.py [hunk 23]
```diff
@@ -4533,3 +4533,107 @@ def test_finish_annotation_file_upload(self, mock_get_aws_creds):
...
+            },
+        )
+        data = {
+            "allowed_email_ids": None,
+        }
+        response = self.client.patch(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data["error"], "Field allowed_email_ids is missing.")
+
+    def test_get_allowed_email_ids_when_challenge_phase_does_not_exist(self):
```
Comment (RishabhJain2018) [inline tests/unit/challenges/test_views.py:4624]: I think there is one test missing i.e. `when a challenge doesn't exist.`

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

File: tests/unit/challenges/test_views.py (hunk 23 above)
Comment (GrimmeryGonsa) [reply tests/unit/challenges/test_views.py:4624]: ah ok.....i am on it..

**Comment by GrimmeryGonsa** [comment]
@RishabhJain2018  it would be great if you could review this PR 😅.....

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**